ironfish-cli/src/commands/service/snapshot.ts

```diff
@@ -0,0 +1,213 @@
+
+    if (flags.path) {
+      exportDir = this.sdk.fileSystem.resolve(flags.path)
+    } else {
+      try {
+        exportDir = await fsAsync.mkdtemp(`${os.tmpdir()}${path.sep}`)
+      } catch (err) {
+        this.log(`Could not create temp folder for snapshot generation`)
+        this.exit(1)
+      }
+    }
+    Assert.isNotUndefined(exportDir)
+
+    const blockExportPath = this.sdk.fileSystem.join(exportDir, 'blocks')
+    await this.sdk.fileSystem.mkdir(blockExportPath, { recursive: true })
+
+    this.log('Connecting to node...')
+
+    const client = await this.sdk.connectRpc()
+
+    const response = client.snapshotChainStream({
+      start: args.start as number | null,
+      stop: args.stop as number | null,
+      maxBlocksPerChunk: flags.maxBlocksPerChunk,
+    })
+
+    const { start, stop } = await AsyncUtils.first(response.contentStream())
+    this.log(`Retrieving blocks from ${start} -> ${stop} for snapshot generation`)
+
+    const progress = CliUx.ux.progress({
+      format: 'Retrieving blocks: [{bar}] {value}/{total} {percentage}% | ETA: {eta}s',
+    }) as ProgressBar
+
+    progress.start(stop - start + 1, 0)
+
+    for await (const result of response.contentStream()) {
+      if (result.buffer && result.seq) {
+        const blockFilePath = this.sdk.fileSystem.join(blockExportPath, `${result.seq}`)
+        await fsAsync.writeFile(blockFilePath, Buffer.from(result.buffer))
+        progress.update(result.seq || 0)
+      }
+    }
+
+    progress.stop()
+
+    const timestamp = Date.now()
+
+    const snapshotFileName = `ironfish_snapshot_${timestamp}.tar.gz`
+    const snapshotPath = this.sdk.fileSystem.join(exportDir, snapshotFileName)
+
+    this.log(`Zipping\n    SRC ${blockExportPath}\n    DST ${snapshotPath}\n`)
+    CliUx.ux.action.start(`Zipping ${blockExportPath}`)
+    await this.zipDir(blockExportPath, snapshotPath)
+    const stat = await fsAsync.stat(snapshotPath)
+    const fileSize = stat.size
+    CliUx.ux.action.stop(`done (${FileUtils.formatFileSize(fileSize)})`)
+
+    const hasher = crypto.createHash('sha256')
+    const fileHandle = await fsAsync.open(snapshotPath, 'r')
+    const stream = fileHandle.createReadStream()
+
+    CliUx.ux.action.start(`Creating checksum for ${snapshotPath}`)
+    for await (const data of stream) {
+      hasher.update(data)
+    }
+    const checksum = hasher.digest().toString('hex')
+    CliUx.ux.action.stop(`done`)
+
+    const blockHeight = stop
+
+    CliUx.ux.action.start(`Uploading to ${bucket}`)
+    await this.uploadToBucket(snapshotPath, bucket, 'application/x-compressed-tar')
+    CliUx.ux.action.stop(`done`)
+
+    await fsAsync.writeFile(
+      path.join(exportDir, 'manifest.json'),
+      JSON.stringify({
+        block_height: blockHeight,
+        checksum,
+        file_name: snapshotFileName,
+        file_size: fileSize,
+        timestamp,
+      }),
+    )
+
+    CliUx.ux.action.start(`Uploading latest snapshot information to ${bucket}`)
```

Comment:
If the write gives an error code for some reason, this would just go ahead and update the manifest and create an invalid state. May be good here to have an `writefile(...).then(...)` in which the then does the manifest upload